File: sdks/python/qnsi/src/qnsi/_activation.py
"""SDK activation — the standard end-user flow.

End-user customers authenticate with a `qnsp_pqc_*` API key issued from the
cloud portal. The SDK calls `/billing/v1/sdk/activate`, which:

1. Validates the API key against the tenant's billing record.
2. Returns the tier name + the limits dict (storage GB, API calls, enclaves
   enabled, etc.) for client-side enforcement.
3. Issues a short-lived activation token cached in memory; subsequent calls
   reuse the same token until ~1 minute before its stated expiry.

This mirrors `@heossi/qnsi-sdk-activation`'s `activateSdk()` in the TypeScript SDKs.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any

import httpx

from qnsi._errors import QnsiApiError, QnsiAuthError, QnsiNetworkError

DEFAULT_BASE_URL = "https://api.qnsi.heossi.com"
ACTIVATION_PATH = "/billing/v1/sdk/activate"
DEFAULT_TIMEOUT_SECONDS = 15.0
EXPIRY_BUFFER_SECONDS = 60.0
SDK_ID = "qnsp-python"
SDK_VERSION = "0.3.0"


@dataclass(frozen=True)
class ActivationResult:
    """Decoded response from the activation handshake."""

    tenant_id: str
    tier: str
    limits: dict[str, Any]
    activation_token: str
    expires_at_epoch: float
    raw: dict[str, Any]


class ApiKeyActivation:
    """Activate a `qnsp_pqc_*` API key against billing-service.

    Concurrency note: this class is NOT thread-safe. Callers that share a
    single `QnsiClient` across threads should rely on the lock that
    `_client.QnsiClient` holds internally; direct users of
    `ApiKeyActivation` should serialize calls themselves.
    """

    def __init__(
        self,
        *,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT_SECONDS,
        http_client: httpx.Client | None = None,
    ) -> None:
        if not api_key or not api_key.strip():
            raise ValueError(
                "api_key is required. Get a free QNSP API key at "
                "https://cloud.qnsi.heossi.com/auth — no credit card required."
            )
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout
        self._http = http_client or httpx.Client(timeout=timeout)
        self._owned_http_client = http_client is None
        self._cached: ActivationResult | None = None

    @property
    def base_url(self) -> str:
        return self._base_url

    @property
    def api_key_header(self) -> dict[str, str]:
        """Return the auth header dict for outbound service calls."""
        return {"authorization": f"Bearer {self._api_key}"}

    def get_activation(self) -> ActivationResult:
        """Return current activation, refreshing if expired or near-expired."""
        cached = self._cached
        now = time.time()
        if cached is not None and cached.expires_at_epoch - EXPIRY_BUFFER_SECONDS > now:
            return cached
        return self._refresh()

    @property
    def tenant_id(self) -> str:
        return self.get_activation().tenant_id

    @property
    def tier(self) -> str:
        return self.get_activation().tier

    @property
    def limits(self) -> dict[str, Any]:
        return self.get_activation().limits

    def has_feature(self, feature: str) -> bool:
        """Convenience: check a billing-side boolean limit (e.g. 'sseEnabled')."""
        return bool(self.get_activation().limits.get(feature, False))

    def invalidate(self) -> None:
        """Drop the cached activation; next call re-mints."""
        self._cached = None

    def close(self) -> None:
        if self._owned_http_client:
            self._http.close()

    def __enter__(self) -> ApiKeyActivation:
        return self

    def __exit__(self, *_exc: object) -> None:
        self.close()

    # ── internal ──────────────────────────────────────────────────────────

    def _refresh(self) -> ActivationResult:
        url = f"{self._base_url}{ACTIVATION_PATH}"
        try:
            response = self._http.post(
                url,
                json={
                    "sdkId": SDK_ID,
                    "sdkVersion": SDK_VERSION,
                    "runtime": "python",
                },
                headers={
                    "authorization": f"Bearer {self._api_key}",
                    "content-type": "application/json",
                },
                timeout=self._timeout,
            )
        except httpx.RequestError as exc:
            raise QnsiNetworkError(
                f"Failed to reach QNSP activation endpoint at {url}: {exc}",
                cause=exc,
            ) from exc

        if response.status_code in (401, 403):
            body = _safe_json(response)
            message = (
                "QNSP rejected your API key. "
                "Get a free API key at https://cloud.qnsi.heossi.com/api-keys"
            )
            if isinstance(body, dict):
                err = body.get("error") or body.get("message")
                if isinstance(err, str):
                    message = err
            raise QnsiAuthError(message, status_code=response.status_code)
        if response.status_code >= 400:
            raise QnsiApiError(
                f"activation failed: {response.text}",
                status_code=response.status_code,
                body=_safe_json(response),
            )

        body = _safe_json(response)
        if not isinstance(body, dict):
            raise QnsiApiError(
                "activation returned an unexpected response shape",
                status_code=response.status_code,
                body=body,
            )

        tenant_id = body.get("tenantId") or body.get("tenant_id")
        tier = body.get("tier") or body.get("planTier") or body.get("plan_tier")
        limits = body.get("limits") or {}
        token = body.get("activationToken") or body.get("activation_token") or body.get("accessToken")
        expires_in = body.get("expiresInSeconds") or body.get("expires_in") or 3600

        if not isinstance(tenant_id, str) or not tenant_id:
            raise QnsiApiError(
                "activation response missing tenantId",
                status_code=response.status_code,
                body=body,
            )

        result = ActivationResult(
            tenant_id=tenant_id,
            tier=str(tier) if isinstance(tier, str) else "unknown",
            limits=limits if isinstance(limits, dict) else {},
            activation_token=token if isinstance(token, str) else "",
            expires_at_epoch=time.time() + float(expires_in),
            raw=body,
        )
        self._cached = result
        return result


def _safe_json(response: httpx.Response) -> object:
    try:
        return response.json()
    except (ValueError, TypeError):
        return None

File: sdks/rust/qnsi/src/activation.rs
//! SDK activation handshake against billing-service.
//!
//! End-user customers authenticate with a `qnsp_pqc_*` API key issued
//! from the cloud portal. The SDK calls `/billing/v1/sdk/activate`,
//! which validates the key, returns the tenant tier + limits, and issues
//! a short-lived activation token cached in memory.

use chrono::{DateTime, Duration, Utc};
use reqwest::StatusCode;
use serde_json::Value;
use std::sync::Mutex;

use crate::errors::{ApiError, AuthError, Error, NetworkError};

const ACTIVATION_PATH: &str = "/billing/v1/sdk/activate";
const EXPIRY_BUFFER_SECONDS: i64 = 60;

/// Decoded response from the activation handshake.
#[derive(Debug, Clone)]
pub struct ActivationResult {
    pub tenant_id: String,
    pub tier: String,
    pub limits: serde_json::Map<String, Value>,
    pub activation_token: String,
    pub expires_at: DateTime<Utc>,
}

/// Performs the SDK activation handshake and caches the result until
/// ~1 minute before expiry.
#[derive(Debug)]
pub struct Activation {
    api_key: String,
    base_url: String,
    http: reqwest::Client,
    cache: Mutex<Option<ActivationResult>>,
}

impl Activation {
    pub(crate) fn new(api_key: String, base_url: String, http: reqwest::Client) -> Self {
        Self {
            api_key,
            base_url: base_url.trim_end_matches('/').to_string(),
            http,
            cache: Mutex::new(None),
        }
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    pub fn auth_header(&self) -> String {
        format!("Bearer {}", self.api_key)
    }

    /// Return the current activation, refreshing if expired or near
    /// expiry. Concurrent callers may both observe a refresh; the result
    /// is only meaningful as a cache hint.
    pub async fn get(&self) -> Result<ActivationResult, Error> {
        if let Some(cached) = self.cache.lock().unwrap().clone() {
            let still_fresh = (cached.expires_at - Utc::now()).num_seconds() > EXPIRY_BUFFER_SECONDS;
            if still_fresh {
                return Ok(cached);
            }
        }
        self.refresh().await
    }

    /// Drop the cached activation. The next call to `get` will trigger a
    /// fresh handshake.
    pub fn invalidate(&self) {
        *self.cache.lock().unwrap() = None;
    }

    async fn refresh(&self) -> Result<ActivationResult, Error> {
        let url = format!("{}{}", self.base_url, ACTIVATION_PATH);
        let body = serde_json::json!({
            "sdkId": crate::SDK_ID,
            "sdkVersion": crate::SDK_VERSION,
            "runtime": "rust",
        });

        let resp = self
            .http
            .post(&url)
            .header("authorization", self.auth_header())
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| NetworkError {
                op: "POST".into(),
                url: url.clone(),
                cause: e.to_string(),
            })?;

        let status = resp.status();
        let raw = resp.text().await.map_err(|e| NetworkError {
            op: "POST".into(),
            url: url.clone(),
            cause: e.to_string(),
        })?;

        if status == StatusCode::UNAUTHORIZED || status == StatusCode::FORBIDDEN {
            return Err(parse_auth_error(status, &raw).into());
        }
        if !status.is_success() {
            return Err(parse_api_error(status, &raw).into());
        }

        let parsed: Value = serde_json::from_str(&raw).map_err(|_| ApiError {
            status_code: status.as_u16(),
            code: None,
            message: "activation response is not valid JSON".into(),
            body: None,
        })?;

        let result = decode_activation(parsed)?;
        *self.cache.lock().unwrap() = Some(result.clone());
        Ok(result)
    }
}

fn decode_activation(raw: Value) -> Result<ActivationResult, ApiError> {
    let obj = raw.as_object().ok_or_else(|| ApiError {
        status_code: 200,
        code: None,
        message: "activation response is not a JSON object".into(),
        body: Some(raw.clone()),
    })?;

    let tenant_id = obj
        .get("tenantId")
        .and_then(|v| v.as_str())
        .ok_or_else(|| ApiError {
            status_code: 200,
            code: None,
            message: "activation response missing tenantId".into(),
            body: Some(raw.clone()),
        })?
        .to_string();

    let tier = obj
        .get("tier")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let limits = obj
        .get("limits")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();

    let activation_token = obj
        .get("activationToken")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let expires_at = obj
        .get("expiresAt")
        .and_then(|v| v.as_str())
        .and_then(|s| DateTime::parse_from_rfc3339(s).ok())
        .map(|dt| dt.with_timezone(&Utc))
        .unwrap_or_else(|| Utc::now() + Duration::minutes(5));

    Ok(ActivationResult {
        tenant_id,
        tier,
        limits,
        activation_token,
        expires_at,
    })
}

pub(crate) fn parse_auth_error(status: StatusCode, raw: &str) -> AuthError {
    let value: Value = serde_json::from_str(raw).unwrap_or(Value::Null);
    let code = value
        .get("code")
        .and_then(|v| v.as_str())
        .map(String::from);
    let message = value
        .get("message")
        .and_then(|v| v.as_str())
        .map(String::from)
        .unwrap_or_else(|| format!("activation rejected (HTTP {})", status.as_u16()));
    AuthError { code, message }
}

pub(crate) fn parse_api_error(status: StatusCode, raw: &str) -> ApiError {
    let value: Value = serde_json::from_str(raw).unwrap_or(Value::Null);
    let code = value
        .get("code")
        .and_then(|v| v.as_str())
        .map(String::from);
    let message = value
        .get("message")
        .or_else(|| value.get("error"))
        .and_then(|v| v.as_str())
        .map(String::from)
        .unwrap_or_else(|| format!("HTTP {}", status.as_u16()));
    ApiError {
        status_code: status.as_u16(),
        code,
        message,
        body: Some(value),
    }
}
